### PR TITLE
[Pytorch] Add source range trace to byte code debug information

### PR DESCRIPTION
**Pull request by cccclai**
## Summary

Follow up https://github.com/pytorch/pytorch/pull/46893, add callstack trace as part of debug information. Since bytecode is not readable, add module full callstack so it's easier to debug.

## test plan
/Users/chenlai/pytorch/cmake-build-debug/bin/test_jit --gtest_filter=LiteInterpreterTest.*:LiteInterpreterTest/*.*:*/LiteInterpreterTest.*/*:*/LiteInterpreterTest/*.* --gtest_color=no

 python test/mobile/test_lite_script_module.py

Example mobile_source_range.pkl
```
(4,
 ('__torch__.C.forward',
  (('module_source_range',
    ('  File "<string>", line 3\n'
     '\n'
     '    def forward(self, x):\n'
     '      return self.A0.forward(self.B0.forward(x))\n'
     '                             ~~~~~~~~~~~~~~~ <--- HERE\n'
     '  \n',
     '  File "<string>", line 3\n'
     '\n'
     '    def forward(self, x):\n'
     '      return self.A0.forward(self.B0.forward(x))\n'
     '             ~~~~~~~~~~~~~~~ <--- HERE\n'
     '  \n')),)))
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48803 [Pytorch] Add source range trace to byte code debug information**

Differential Revision: [D25712318](https://our.internmc.facebook.com/intern/diff/D25712318)

### CODE REVIEW

**Comment by cccclai** [comment]
This change is still in draft. I'd like to know if it's in the right direction.

File: torch/csrc/jit/serialization/export_module.cpp [hunk 1]
```diff
@@ -54,6 +54,17 @@ static IValue Table(
   return Tup(std::move(ivalue_entries));
 }
 
+std::string getSourceRangeTrace(
+    Node* node,
+    const std::string& root_scope_string) {
+  std::string source_range_trace;
+  if (node->callstack()) {
+    auto callstack_ptr = *(node->callstack());
+    source_range_trace = callstack_ptr->source_range_trace();
```
Comment (iseeyuan) [inline torch/csrc/jit/serialization/export_module.cpp:63]: Why not we directly get the source range, instead of the source_range_trace?

Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:63]: The naming is probably a bit confusing. Each node has it's own callstack, and source range is the callstack function entry. `source_range_trace` means the whole callstack. For example, for the test case `TEST(LiteInterpreterTest, HierarchyModuleInfo) ` in `test/cpp/jit/test_lite_interpreter.cpp`:

 if source_range is used, the result will be

```
{
  "top(C).forward{}",
  "top(C).B0(B).forward{self.B0.forward:}",
  "top(C).B0(B).forward.A0(A).forward{self.A0.forward:}"}
);
```

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Some general thoughts:
- I am of the opinion that we should store source_range information per node in a separate source_debug_info or some sort table. This is because module debug info is generally used for profiling so that you can find populate module level latency using operator level info. Adding source range information embedded in there makes annotation harder and I think it does not belong there.
- Probably add a separate option (or maybe not) for saving this, independent of module_debug_info. But I am not too hung up on this.
- Also the part where source range information is going to be most useful is when source range contains information from python source code. So I would say for your tests cases python tests would help more.

File: torch/csrc/jit/serialization/export_module.cpp (hunk 1 above)
Comment (iseeyuan) [reply torch/csrc/jit/serialization/export_module.cpp:63]: Maybe the goal is ambitious, but I'm looking forward to something like this:

```
  File "/test/mobile#binary,link-tree/mobile/test_lite_script_module_namedtuple.py", line 341
            def forward(self, x: int):
                p = myNamedTuple(x)
                    ~~~~~~~~~~~~ <--- HERE
                return p
```
Or something similar that could highlight the original Python code.

**Comment by iseeyuan** [comment]
> Some general thoughts:
> 
> * I am of the opinion that we should store source_range information per node in a separate source_debug_info or some sort table. This is because module debug info is generally used for profiling so that you can find populate module level latency using operator level info. Adding source range information embedded in there makes annotation harder and I think it does not belong there.
> * Probably add a separate option (or maybe not) for saving this, independent of module_debug_info. But I am not too hung up on this.
> * Also the part where source range information is going to be most useful is when source range contains information from python source code. So I would say for your tests cases python tests would help more.

Agree on the separate session idea. The source range is for general debugging, independent of profiling.

File: torch/csrc/jit/serialization/export_module.cpp (hunk 1 above)
Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:63]: Ah that should be an easy change. I was worried that the size will explode, so I only include the call function name. Do you think it's a concern, or it should be fine to do some pretty print?

Comment (kimishpatel) [reply torch/csrc/jit/serialization/export_module.cpp:63]: I agree with Martin. This should be doable. Plus, doing this in a separate table can help selectively save either module debug info or source debug info. As with other things, such as compiling code with debug symbols, we expect to pay for any increase in code size for ease of debug.

**Pull request closed without merging**